### PR TITLE
docs: developer guidance on adding dependencies

### PR DESCRIPTION
**Pull request by leongdl**
### What was the problem/requirement? (What/Why)
- It is not clear what deadline cloud's guidance was to using 3rd party dependencies.
 
### What was the solution? (How)
- Adding documentation and expectations on PRs with new dependencies.
- Update Pull request template, and also `DEVELOPMENT.md`.

### What is the impact of this change?
- Documentation, improve developer experience.

### How was this change tested?
NA.

- Have you run the unit tests?
NA

- Have you run the integration tests?
NA

### Was this change documented?
- This PR is documentation.

### Is this a breaking change?
- No.

### Does this change impact security?
- No

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -20,6 +20,12 @@ See [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainlin
 - Are relevant docstrings in the code base updated?
 - Has the README.md been updated? If you modified CLI arguments, for instance.
 
+### Does this PR introduce new dependencies?
+
+- Does the PR add new package dependencies to `dependencies` section of `pyproject.toml`?
+- Please consider why a new dependency is needed? Dependency libraries need to be updated or maintained.
```
Comment (crowecawcaw) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- Please consider why a new dependency is needed. Dependency libraries need to be updated or maintained.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -20,6 +20,12 @@ See [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainlin
 - Are relevant docstrings in the code base updated?
 - Has the README.md been updated? If you modified CLI arguments, for instance.
 
+### Does this PR introduce new dependencies?
+
+- Does the PR add new package dependencies to `dependencies` section of `pyproject.toml`?
```
Comment (mwiebe) [inline .github/PULL_REQUEST_TEMPLATE.md]: A short description of why it's better to minimize dependencies could help with reasoning about it. The main reason is that this library is designed to be integrated into third-party applications by customers who have highly customized deployments. Every additional dependency increases the risk of version conflicts with something they've already set up.

File: DEVELOPMENT.md [hunk 3]
```diff
@@ -201,6 +201,13 @@ For the Python library interface:
 * Changing the location that a file or directory is created should be considered to be a breaking change. These locations have a tendancy to become
   de-facto parts of the public contract as users build automation that assumes these locations is unchanged.
 
+### Library Dependencies
```
Comment (jusiskin) [inline DEVELOPMENT.md:205]: Other things that should be considered when adding new dependencies:

### quality of the dependency

- Is the dependency reputable or maintained by a reputable source? Not sure how we can quantify this, but some ideas include:
    -  PyPI download stats?
    - GitHub stars?
    - GitHub dependency graph showing downstream consumers?
- Is it well-maintained? does it release regularly/recently?

### version pinning

How should we pin this new dependency? Does the dependency document their versioning scheme (e.g. semver)?

`deadline-cloud` is both a python library and an application in the case of the CLI. Libraries should strive to accommodate a range of versions to allow consumers to have some flexibility in their closures. Applications should be more opinionated to enforce more reproducible behavior.

In the current structure, we need to balance between both. This is probably a "use your best judgement" decision, but we should prompt the contributor to think about this.

### licensing

*   We also need to ensure the license of the dependency meets our needs
*   We also aim to attribute our dependencies in https://github.com/aws-deadline/deadline-cloud/blob/mainline/THIRD_PARTY_LICENSES

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -20,6 +20,12 @@ See [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainlin
 - Are relevant docstrings in the code base updated?
 - Has the README.md been updated? If you modified CLI arguments, for instance.
 
+### Does this PR introduce new dependencies?
```
Comment (jusiskin) [inline .github/PULL_REQUEST_TEMPLATE.md:23]: Another option is to have two checkboxes:

*   [ ] This PR adds one or more new dependency Python packages. I acknowledge I have reviewed the considerations for adding dependencies in `DEVELOPMENT.md`.
*   [ ] This PR does not add any new dependencies.

**Review by jusiskin — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (leongdl) [reply .github/PULL_REQUEST_TEMPLATE.md]: Sure thing, let me change

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (leongdl) [reply .github/PULL_REQUEST_TEMPLATE.md]: Nice! I'll add a framing into here.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 4 above)
Comment (leongdl) [reply .github/PULL_REQUEST_TEMPLATE.md:23]: NIce! I like the checkbox idea.

File: DEVELOPMENT.md (hunk 3 above)
Comment (leongdl) [reply DEVELOPMENT.md:205]: Totally agree - these are all good things to consider when selecting 3rd party libs.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 4 above)
Comment (leongdl) [reply .github/PULL_REQUEST_TEMPLATE.md:23]: Adopted this checkbox idea, also changed the link to `DEVELOPMENT.md`

File: DEVELOPMENT.md (hunk 3 above)
Comment (leongdl) [reply DEVELOPMENT.md:205]: Added - I reframed the sections The Version pinning one I think needs more work.

**Review by crowecawcaw — APPROVED** [review]

**Review by epmog — APPROVED** [review]

**Review by jusiskin — APPROVED** [review]

**Pull request merged**